src/safeposix/syscalls/fs_calls.rs

```diff
@@ -1844,4 +1845,100 @@ impl Cage {
     }
+
+    //------------------SHMGET SYSCALL------------------
+
+    pub fn shmget_syscall(&self, key: i32, size: usize, shmflg: i32)-> i32 {
+        if key == IPC_PRIVATE {return syscall_error(Errno::ENOENT, "shmget", "IPC_PRIVATE not implemented");}
+        if (size as i32) < SHMMIN || (size as i32) > SHMMAX { return syscall_error(Errno::EINVAL, "shmget", "Size is less than SHMMIN or more than SHMMAX"); }
+        let shmid: i32;
+        let metadata = &SHM_METADATA;
+
+        if metadata.shmkeyidtable.contains_key(&key) {
+            if (IPC_CREAT | IPC_EXCL) == (shmflg & (IPC_CREAT | IPC_EXCL)) {
+                return syscall_error(Errno::EEXIST, "shmget", "key already exists and IPC_CREAT and IPC_EXCL were used");
+            }
+            shmid = *metadata.shmkeyidtable.get(&key).unwrap(); 
+        } else {
+            if 0 == (shmflg & IPC_CREAT) {
+                return syscall_error(Errno::ENOENT, "shmget", "tried to use a key did not exist, and IPC_CREAT was not specified");
+            }
+            shmid = metadata.new_keyid(key);
+            let mode = (shmflg & 0x1FF) as u16; // mode is 9 least signficant bits of shmflag, even if we dont really do anything with them
+
+            let segment = new_shm_segment(key, size, self.cageid as i32, DEFAULT_UID, DEFAULT_GID, mode);
+            metadata.shmtable.insert(shmid, segment);
+        }
+
+        shmid // return the shmid
+    }
+
+    //------------------SHMAT SYSCALL------------------
+
+    pub fn shmat_syscall(&self, shmid: i32, shmaddr: *mut u8, shmflg: i32)-> i32 {
+        let metadata = &SHM_METADATA;
+        let prot : i32;
+        if let Some(mut segment) = metadata.shmtable.get_mut(&shmid) {
+
+            if 0 == (shmflg & SHM_RDONLY) {
+                prot = PROT_READ;
+            }  else { prot = PROT_READ | PROT_WRITE; }
+
+            segment.map_shm(shmaddr, prot);
+            metadata.rev_shm_add(self.cageid as i32, shmaddr, shmid);
+        } else { return syscall_error(Errno::EINVAL, "shmat", "Invalid shmid value"); }
+
+        0 //shmat has succeeded!
+    }
+
+    //------------------SHMDT SYSCALL------------------
+
+    pub fn shmdt_syscall(&self, shmaddr: *mut u8)-> i32 {
+        let metadata = &SHM_METADATA;
+        let mut rm = false;
+        if let Some(shmid) = metadata.rev_shm_lookup(self.cageid as i32, shmaddr) {
+
+            let mut segment = metadata.shmtable.get_mut(&shmid).unwrap();
+
+            segment.unmap_shm(shmaddr);
+    
+            if segment.rmid && !segment.shminfo.shm_nattach == 0 { rm = true; }           
+            metadata.rev_shm_rm(self.cageid as i32, shmaddr);
+    
+            if rm {
+                let key = segment.key;
+                drop(segment);
+    
+                metadata.shmtable.remove(&shmid);
+                metadata.shmkeyidtable.remove(&key);
+            }
+
+        } else { return syscall_error(Errno::EINVAL, "shmdt", "No shared memory segment at shmaddr"); }
```

Comment:
I think there might be race conditions with shmat and shmdt relating to nattach. I think you may have to wrap shm_nattach in a lock, and lock around shmat/shmdt to ensure things?
This bears further looking into.

Reply:
Shouldn't having dashmap references, both to shm metadata and the info struct deal with concurrency here?

Reply:
But we don't have the dashmap reference after segment is dropped, also you don't have a reference to the shmtable because you just have an i32